sklearn/base.py

```diff
@@ -246,10 +246,29 @@ def __repr__(self):
 
         repr_ = pp.pformat(self)
 
-        # Use bruteforce ellipsis if string is very long
+        # Use bruteforce ellipsis if string is very long.
+        # We will keep (approximately) N_CHAR_MAX non-blank characters.
         if len(''.join(repr_.split())) > N_CHAR_MAX:  # check non-blank chars
-            lim = N_CHAR_MAX // 2
-            repr_ = repr_[:lim] + '...' + repr_[-lim:]
+            lim = N_CHAR_MAX // 2  # apprx number of chars to keep on both ends
+            repr_array = np.array(list(repr_))
```

Comment:
That's the only way I found :/

```ipython
In [2]: np.array('abcdef')
Out[2]: array('abcdef', dtype='<U6')

In [3]: np.array(list('abcdef'))
Out[3]: array(['a', 'b', 'c', 'd', 'e', 'f'], dtype='<U1')

In [7]: np.array('abcdef', dtype='<U1')
Out[7]: array('a', dtype='<U1')
```


I'm also not a big fan of using numpy here TBH. What we ultimately need is to find the nth non-blank character in the string, from the start and from the end. I couldn't think of a pythonic way to do this, but I'd be happy to consider alternative solutions!

------

I don't think we should constraint the number of lines because lots of lines doesn't necessarily mean lots of non-blank characters.

Are you suggesting we could constraint the number of non blank characters, and cut according to the number of lines, something like (warning, buggy):

```python
        if len(''.join(repr_.split())) > N_CHAR_MAX:  # check non-blank chars
            N_LINES = 10  # totally arbitrary
            splitted = repr_.split('\n')
            splitted = splitted[:N_LINES] + ['...'] + splitted[-N_LINES:]
            repr_ = '\n'.join(splitted)
```

I would find it reasonable. But it might be hard to find an appropriate value for `N_LINES` so that the number of non-blank characters is approximately equal to `N_CHAR_MAX`.

Reply:
```py
np.array('abcdef').view(('<U1', 6))
```
does what you want in numpy

```py
def nth_nonblank(s, n):
    return re.match(r'^(\s*\S){%d}' % n, s).end() - 1

%```
does what you want in regex.

```py
def nth_nonblank(s, n):
    return next(itertools.islice((i for i, c in enumerate(s)
                                  if c not in string.whitespace),
                                 n - 1, n))
```
does what you want in iterators.

Okay. Let's not worry about the number of lines constraint for now.

Reply:
Wow impressive... Thanks!!

I went for the regex solution which i believe is fairly easy to understand, given some appropriate comments.